x/mint/types/minter.go

```diff
@@ -0,0 +1,76 @@
+package types
+
+import (
+	"fmt"
+	"time"
+
+	"cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// NewMinter returns a new Minter object.
+func NewMinter(inflationRate sdk.Dec, annualProvisions sdk.Dec, genesisTime *time.Time, bondDenom string) Minter {
+	return Minter{
+		InflationRate:    inflationRate,
+		AnnualProvisions: annualProvisions,
+		GenesisTime:      genesisTime,
+		BondDenom:        bondDenom,
+	}
+}
+
+// DefaultMinter returns a Minter object with default values.
+func DefaultMinter() Minter {
+	unixEpoch := time.Unix(0, 0).UTC()
+	return NewMinter(initalInflationRate, sdk.NewDec(0), &unixEpoch, sdk.DefaultBondDenom)
+}
+
+// Validate returns an error if the minter is invalid.
+func (m Minter) Validate() error {
+	if m.InflationRate.IsNegative() {
+		return fmt.Errorf("inflation rate %v should be positive", m.InflationRate.String())
+	}
+	if m.AnnualProvisions.IsNegative() {
+		return fmt.Errorf("annual provisions %v should be positive", m.AnnualProvisions.String())
+	}
+	if m.BondDenom == "" {
+		return fmt.Errorf("bond denom should not be empty string")
+	}
+	return nil
+}
+
+// CalculateInflationRate returns the inflation rate for the current year depending on
+// the current block height in context. The inflation rate is expected to
+// decrease every year according to the schedule specified in the README.
+func (m Minter) CalculateInflationRate(ctx sdk.Context) sdk.Dec {
+	years := yearsSinceGenesis(*m.GenesisTime, ctx.BlockTime())
+	inflationRate := initalInflationRate.Mul(sdk.OneDec().Sub(disinflationRate).Power(years))
+
+	if inflationRate.LT(targetInflationRate) {
+		return targetInflationRate
+	}
+	return inflationRate
+}
+
+// CalculateAnnualProvisions returns the total number of tokens that should be
+// minted due to inflation for the current year.
+func (m Minter) CalculateAnnualProvisions(totalSupply math.Int) sdk.Dec {
+	return m.InflationRate.MulInt(totalSupply)
+}
+
+// CalculateBlockProvision returns the total number of coins that should be
+// minted due to inflation for the current block.
+func (m Minter) CalculateBlockProvision(current time.Time, previous time.Time) sdk.Coin {
+	timeElapsed := current.Sub(previous).Seconds()
+	portionOfYear := sdk.NewDec(int64(timeElapsed)).Quo(sdk.NewDec(int64(SecondsPerYear)))
+	blockProvision := m.AnnualProvisions.Mul(portionOfYear)
+	return sdk.NewCoin(m.BondDenom, blockProvision.TruncateInt())
+}
```

Comment:
Do you know what the error of this function is? i.e. the difference between the actual `AnnualProvision` and multiplying this over x many blocks with ranging intervals around 15 seconds.

Reply:
it looks pretty close:

```go
func TestCalculateBlockProvisionError(t *testing.T) {
	minter := DefaultMinter()
	current := time.Date(2023, 1, 2, 0, 0, 0, 0, time.UTC)
	end := time.Date(2024, 1, 2, 0, 0, 0, 0, time.UTC)

	totalSupply := sdk.NewDec(1_000_000_000_000)             // 1 trillion utia
	annualProvisions := totalSupply.Mul(initalInflationRate) // 80 billion utia

	minter.AnnualProvisions = annualProvisions
	totalBlockProvisions := sdk.NewDec(0)
	for current.Before(end) {
		blockInterval := randomBlockInterval()
		previous := current
		current = current.Add(blockInterval)
		got := minter.CalculateBlockProvision(current, previous)
		totalBlockProvisions = totalBlockProvisions.Add(sdk.NewDecFromInt(got.Amount))
	}
	fmt.Printf("totalBlockProvisions %v\n", totalBlockProvisions)
	fmt.Printf("annualProvisions %v\n", annualProvisions)
}

func randomBlockInterval() time.Duration {
	return time.Duration(randInRange(14, 16)) * time.Second
}

// randInRange returns a random number in the range (min, max) inclusive.
func randInRange(min int, max int) int {
	return rand.Intn(max-min) + min
}
```

gives the output

```
totalBlockProvisions 79945957690.000000000000000000
annualProvisions 80000000000.000000000000000000
```

Reply:
Ok one last thing. Can you try the same but instead of 

```go
func randomBlockInterval() time.Duration {
	return time.Duration(randInRange(14, 16)) * time.Second
}
```
use
```
func randomBlockInterval() time.Duration {
	return time.Duration(randInRange(14000, 16000)) * time.Millisecond
}
```

Reply:
```
totalBlockProvisions 77283450905.000000000000000000
annualProvisions 80000000000.000000000000000000
```